File: Data Structures & Algorithms/search-2d-matrix/submission-1.ts
class Solution {
    /**
     * @param {number[][]} matrix
     * @param {number} target
     * @return {boolean}
     */
    searchMatrix(matrix: number[][], target: number): boolean {
        let low = 0;
        let high = (matrix[0].length * matrix.length) - 1;
        let mid;

        while(low <= high) {
            mid = low + Math.floor((high - low)/2);

            const rowIdx = Math.floor(mid / matrix[0].length);
            const colIdx = mid % matrix[0].length;
            const element = matrix[rowIdx][colIdx];
            //console.log("indices = ", low, mid, high, `matrix[${rowIdx}][${colIdx}]`, element);
            if (element === target) {
                return true;
            }
            if (target > element) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return false;
    }
}

File: Data Structures & Algorithms/binary-search/submission-1.java
class Solution {
    public int search(int[] nums, int target) {
        int result = -1;

        int low = 0;
        int high = nums.length - 1;

        while (low <= high) {
            int index = low + ((high - low) / 2);
            int currentValue = nums[index];

            if (currentValue == target) {
                return index;
            }
            if (target < currentValue) {
                high = index - 1;
            } else {
                low = index + 1;
            }
            
        }
        return result;
    }
}
